### PR TITLE
Add Management Command for Data Purge

### PR DESCRIPTION
**Pull request by mkangia**

### CODE REVIEW

**Comment by mkangia** [comment]
@czue : I might have pushed this too far. Share reviews please.

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+from django.core.management.base import BaseCommand, CommandError
+from corehq.apps.app_manager.models import Application
+from casexml.apps.case.xform import get_case_ids_from_form
+from corehq.form_processor.interfaces.dbaccessors import FormAccessors, CaseAccessors
+from dimagi.utils.django.management import are_you_sure
+
+
+X_FORM_HEADER = ['Version', 'Form Id', 'Domain']
+CASE_HEADER = ['CaseID', 'CaseType', 'CaseOpenedOn', 'CaseName', 'CaseUserId', 'CaseXFormIds']
+X_FORM_FILENAME = 'X-Form-Details.csv'
```
Comment (czue) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: super nit picky but "XForm" would be more consistent with naming everywhere else

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+CASE_FILE_NAME = 'Case-Details.csv'
+
+
+# Expects 4 arguments in order : domain app_id version_number test_run
+# ex: ./manage.py purge_forms_and_cases testapp c531daeece0633738c9a3676a13e3d4f 88 yes
+# domain is included with app_id to ensure the user knows what app to delete
+# version_number to delete data accumulated by versions BEFORE this : integer
+# test_run should be yes(case-sensitive) for a test_run and anything otherwise
+# though deletion would be re-confirmed so no panic
```
Comment (czue) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: you can add this as an attribute on the class called `help` and then it will be be printed out when someone runs `./manage.py purge_forms_and_cases --help`

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        self.case_ids = set()
+        self.filtered_xform_ids, self.xform_ids = [], []
+        self.xform_writer, self.case_writer = None, None
+        self.forms_accessor, self.case_accessors = None, None
+        self.domain, self.app_id, self.version_number, self.test_run = None, None, None, None
+
+    def setup(self):
+        self.xform_writer = csv.writer(open(X_FORM_FILENAME, 'w+b'))
+        self.xform_writer.writerow(X_FORM_HEADER)
+        self.case_writer = csv.writer(open(CASE_FILE_NAME, 'w+b'))
```
Comment (czue) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: might want to add a timestamp to the filenames in case this gets run multiple times so it doesn't blow stuff away

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        self.iterate_forms_and_collect_case_ids()
+        _print_final_debug_info(self.xform_ids, self.filtered_xform_ids, self.case_ids)
+        if self.data_to_delete() and self.delete_permitted():
+            self.delete_forms_and_cases()
+            print "Process Completed!! Keep copy of files %s, %s" % (X_FORM_FILENAME, CASE_FILE_NAME)
+        else:
+            print 'Process Finished w/o Changes..'
+
+    def iterate_forms_and_collect_case_ids(self):
+        print 'Iterating Through X-Forms and Collecting Case Ids'
```
Comment (czue) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: could be nice to say how many it's doing here.

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        return len(self.filtered_xform_ids) != 0 or len(self.case_ids) != 0
+
+    def delete_forms_and_cases(self):
+        print 'Proceeding with deleting forms and cases'
+        self.forms_accessor.soft_delete_forms(self.filtered_xform_ids)
+        self.case_accessors.soft_delete_cases(list(self.case_ids))
+
+
+def _print_form_details(xform, file_writer):
+    print X_FORM_HEADER
```
Comment (czue) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: should the headers not also be in the files?

**Comment by czue** [comment]
nice this looks really good and i like how you broke everything up. had a couple small comments but nothing too big or blocking

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+XFORM_HEADER = ['Version', 'Form Id', 'Domain']
+CASE_HEADER = ['CaseID', 'CaseType', 'CaseOpenedOn', 'CaseName', 'CaseUserId', 'CaseXFormIds']
+CURRENT_TIME = datetime.strftime(datetime.now(), '-%Y-%m-%d-%H:%M:%S')
+XFORM_FILENAME = "XForm-Details%s.csv" % CURRENT_TIME
+CASE_FILE_NAME = 'Case-Details%s.csv' % CURRENT_TIME
+
+
+
+class Command(BaseCommand):
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: E303 too many blank lines (3)

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py (hunk 5 above)
Comment (mkangia) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: They are .. its a part of the setup itself.

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py (hunk 4 above)
Comment (mkangia) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: Cool.

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py (hunk 3 above)
Comment (mkangia) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: Absolutely. Added.
Is there a better place to put these files like a tmp or data dir we use?

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py (hunk 2 above)
Comment (mkangia) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: Nice!! 👍

Comment (mkangia) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: Also,
as a fallback to revert the changes will
> In Case of bad changes revert doc_type to CommCareCase and XformInstance using ids

work? Was using that while re-running script for debugging. Would be good to add a way out in "help" itself if there is one?

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            self.delete_forms_and_cases()
+            print "Process Completed!! Keep copy of files %s, %s" % (X_FORM_FILENAME, CASE_FILE_NAME)
+        else:
+            print 'Process Finished w/o Changes..'
+
+    def iterate_forms_and_collect_case_ids(self):
+        print 'Iterating Through X-Forms and Collecting Case Ids'
+        for xform in self.forms_accessor.iter_forms(self.xform_ids):
+            _print_form_details(xform, self.xform_writer)
+            if '@version' in xform.form and int(xform.form['@version']) < self.version_number:
```
Comment (czue) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: because of the way form versioning works this isn't going to be the same as the app version. Have to use `Application.get(form.build_id).version` instead (and should probably keep a cache of this to avoid multiple lookups per app)

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py [hunk 8]
```diff
@@ -78,6 +82,13 @@ def iterate_forms_and_collect_case_ids(self):
         if self.case_ids:
             self.print_case_details()
 
+    def get_xform_build_version(self, xform):
+        version_from_mapping = self.version_mapping.get(xform.build_id)
+        if not version_from_mapping:
+            version_from_mapping = int(get_app(self.domain, xform.build_id).version)
```
Comment (czue) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: I think this should also check if `xform.build_id` is `None` and/or handle failures in `get_app` not returning anything since that isn't guaranteed to be set (e.g. if the form was submitted from case importer or an API)

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py [hunk 9]
```diff
@@ -68,8 +69,11 @@ def handle(self, *args, **options):
     def iterate_forms_and_collect_case_ids(self):
         print "Iterating Through %s XForms and Collecting Case Ids" % len(self.xform_ids)
         for xform in self.forms_accessor.iter_forms(self.xform_ids):
+            # Get app version by fetching app corresponding to xform build_id since xform.form
+            # does not have updated app version unless form was updated for that version
+            xform.app_version_built_with = self.get_xform_build_version(xform)
```
Comment (czue) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: it feels a little funny to me to patching this property onto the form if you can just pass it through to `_form_details` instead

Comment (mkangia) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: Not sure. The purpose is to bind it to the xform object and easy to refer when needed. May be if you can share the exact change i might see it better.

Comment (czue) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: the danger of adding an arbitrary property to the root object is that because of the way couch `Document`s work if someone later called `form.save()` it would persist that property to the database which seems like it would be an unintended side effect of this code. I don't think it is a blocker but in this instance I think you can define `app_version_built_with` as a variable and then just pass it through to the other calls where you need it to avoid that potential problem

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py (hunk 8 above)
Comment (mkangia) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: Added a check for build_id and then also in case app is not found. Should cover possible edge cases though not tested.

File: corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py (hunk 9 above)
Comment (mkangia) [inline corehq/ex-submodules/couchforms/management/commands/purge_forms_and_cases.py]: Oh! Thats dangerous! Ouch Couch! :hurtrealbad: 
Does not seem to be a blocker but would still make the change to make this a habit.

**Comment by czue** [comment]
@mkangia I think this is g2g from my side but not merging due to label

**Comment by czue** [comment]
(in the future feel free to remove label whenever you think it's safe to merge)

**Comment by mkangia** [comment]
@czue : Sorry, kept the label to ensure to wrap up the change for "_form_details" variable change.
